core/aws-core/src/main/java/software/amazon/awssdk/awscore/interceptor/TraceIdExecutionInterceptor.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.awscore.interceptor;
+
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.SystemSetting;
+
+/**
+ * The {@code TraceIdExecutionInterceptor} copies the {@link #TRACE_ID_ENVIRONMENT_VARIABLE} value to the
+ * {@link #TRACE_ID_HEADER} header, assuming we seem to be running in a lambda environment.
+ */
+@SdkInternalApi
+public class TraceIdExecutionInterceptor implements ExecutionInterceptor {
+    private static final String TRACE_ID_HEADER = "X-Amzn-Trace-Id";
+    private static final String TRACE_ID_ENVIRONMENT_VARIABLE = "_X_AMZN_TRACE_ID";
+    private static final String LAMBDA_FUNCTION_NAME_ENVIRONMENT_VARIABLE = "AWS_LAMBDA_FUNCTION_NAME";
+
+    @Override
+    public SdkHttpRequest modifyHttpRequest(Context.ModifyHttpRequest context, ExecutionAttributes executionAttributes) {
+        Optional<String> traceIdHeader = traceIdHeader(context);
+        if (!traceIdHeader.isPresent()) {
+            Optional<String> lambdafunctionName = lambdaFunctionNameEnvironmentVariable();
+            Optional<String> traceId = traceIdEnvironmentVariable();
+
+            if (lambdafunctionName.isPresent() && traceId.isPresent()) {
+                return context.httpRequest().copy(r -> r.putHeader(TRACE_ID_HEADER, traceId.get()));
+            }
+        }
```

Comment:
Our HTTP clients should encode invalid characters automatically, shouldn't they? Let me check...

Reply:
It looks like they're validated and quoted, but not URL encoded, in the SDK's HTTP clients. Let me confirm whether that's okay with the spec writer.

Reply:
Nice thanks for checking!

Reply:
Confirmed that all that matters is the input not being able to corrupt the request. So percent encoding isn't required, in our case, since our HTTP clients handle it appropriately.

Reply:
Can we make sure it works Netty client works? Our support for escaping headers might not be as good there